File: solutions/permutations/permutations.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> pre, cur;
        for (const auto& num : nums) {
            if (pre.size() == 0) {
                pre.push_back({num});
            } else {
                for (auto permutation : pre) {
                    int size = permutation.size();
                    for (int i = 0; i <= size; i++) {
                        permutation.insert(permutation.begin() + i, num);
                        cur.push_back(permutation);
                        permutation.erase(permutation.begin() + i);
                    }
                }
                pre = cur;
                cur.clear();
            }
        }
        return pre;
    }
};

int main() {
    Solution solver;
    vector<int> nums = {1, 2, 3};
    auto result = solver.permute(nums);
    for (const auto& permutation : result) {
        for (const auto& num : permutation) {
            cout << num << " ";
        }
        cout << endl;
    }
    return 0;
}

File: solutions/permutations/permutations.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.dfs(nums, [], result)
        return result
    
    def dfs(self, nums, path, result):
        
        if len(nums)==0:
            result.append(path)
        else:
            for i in range(len(nums)):
                self.dfs(nums[:i]+nums[i+1:], path+[nums[i]], result)

        
        
if __name__ == "__main__":
    solver = Solution()
    nums = [1,2,3]
    print(solver.permute(nums))

